File: dtr-bench/custom_dtr/sepsis_types.py
from typing import List, Any, Optional
from collections import namedtuple
import random
from enum import Enum
import numpy as np

Action = namedtuple('Action', ['abx', 'vaso', 'vent'])

State = namedtuple('State', ['hr', 'bp', 'o2', 'glu',
                   'diabetic', 'abx', 'vaso', 'vent'])


EnvParameters = namedtuple('EnvParameters', [
    'abx_on_hr_H_N',
    'abx_on_bp_H_N',
    'abx_withdrawn_hr_N_H',
    'abx_withdrawn_bp_N_H',
    'vent_on_o2_L_N',
    'vent_withdrawn_o2_N_L',
    'nond_vaso_on_bp_L_N',
    'nond_vaso_on_bp_N_H',
    'diab_vaso_on_bp_L_N',
    'diab_vaso_on_bp_L_H',
    'diab_vaso_on_bp_N_H',
    'diab_vaso_on_glu_up',
    'nond_vaso_withdrawn_bp_N_L',
    'nond_vaso_withdrawn_bp_H_N',
    'diab_vaso_withdrawn_bp_N_L',
    'diab_vaso_withdrawn_bp_H_N',
    'fluct',
    'diab_fluct_glu'
])


class Level(Enum):
    LOW = -1
    NORMAL = 0
    HIGH = 1
    SUPER_LOW = -2
    SUPER_HIGH = 2


ACTIONS = [
    Action(abx, vaso, vent)
    for abx in [True, False]
    for vaso in [True, False]
    for vent in [True, False]
]

STATES = [
    State(hr, bp, o2, glu, diabetic, abx, vaso, vent)
    for hr in [Level.LOW, Level.NORMAL, Level.HIGH]
    for bp in [Level.LOW, Level.NORMAL, Level.HIGH]
    for o2 in [Level.LOW, Level.NORMAL]
    for glu in [Level.SUPER_LOW, Level.LOW, Level.NORMAL, Level.HIGH, Level.SUPER_HIGH]
    for diabetic in [True, False]
    for abx in [True, False]
    for vaso in [True, False]
    for vent in [True, False]
]


class Episode:
    def __init__(self,
                 policy: Optional[List[int]] = None,
                 rewards: List[float] = None,
                 visited: Optional[List[int]] = None):
        # Default to an empty list if None
        self.policy: Optional[List[int]] = policy
        # Default to an empty list if None
        self.rewards: List[float] = rewards or []
        # Default to an empty list if None
        self.visited: Optional[List[int]] = visited or []
        self.date: str = str(np.datetime64('now'))


state_to_index = {state: i for i, state in enumerate(STATES)}
action_to_index = {action: i for i, action in enumerate(ACTIONS)}

Policy = List[int]  # for each state, index of the action to take


def random_policy():
    return [action_to_index[random.choice(ACTIONS)] for _ in STATES]


def random_initial_state():
    # returns a random initial state with all actions set to False

    return State(
        Level(random.randint(-1, 1)),
        Level(random.randint(-1, 1)),
        Level(random.randint(-1, 0)),
        Level(random.randint(-2, 2)),
        random.choice([True, False]),
        False,
        False,
        False
    )

File: sepsis/custom_sepsis/src/custom_sepsis/models/ppl/model/sepsis_types.jl
module SepsisTypes
export Action, Level, SUPER_LOW, LOW, NORMAL, HIGH, SUPER_HIGH, State, Policy, get_env_params, STATES, ACTIONS, to_state, to_action, to_policy, HR_LEVELS, BP_LEVELS, O2_LEVELS, GLU_LEVELS, SepsisParams, Policy, to_gym_pol

struct Action
    abx::Bool
    vaso::Bool
    vent::Bool
end

@enum Level begin
    SUPER_LOW = -2
    LOW = -1
    NORMAL = 0
    HIGH = 1
    SUPER_HIGH = 2
end

HR_LEVELS = [LOW, NORMAL, HIGH]
BP_LEVELS = [LOW, NORMAL, HIGH]
O2_LEVELS = [LOW, NORMAL]
GLU_LEVELS = [SUPER_LOW, LOW, NORMAL, HIGH, SUPER_HIGH]

struct State
    hr::Level
    bp::Level
    o2::Level
    glu::Level
    diabetic::Bool
    abx::Bool
    vaso::Bool
    vent::Bool

    function State(hr::Level, bp::Level, o2::Level, glu::Level, diabetic::Bool, abx::Bool, vaso::Bool, vent::Bool)
        # Validate hr and bp (only LOW, NORMAL, HIGH are allowed)
        @assert hr in HR_LEVELS "hr must be LOW, NORMAL, or HIGH"
        @assert bp in BP_LEVELS "bp must be LOW, NORMAL, or HIGH"

        # Validate o2 (only LOW, NORMAL are allowed)
        @assert o2 in O2_LEVELS "o2 must be LOW or NORMAL"
        @assert glu in GLU_LEVELS

        # No restrictions for glu; it can take any Level
        return new(hr, bp, o2, glu, diabetic, abx, vaso, vent)
    end
end

function to_state(state::Tuple)::State
    hr, bp, o2, glu, diabetic, abx, vaso, vent = state
    return State(Level(hr), Level(bp), Level(o2), Level(glu), diabetic, abx, vaso, vent)
end
function to_action(action::Tuple)::Action
    return Action(action...)
end
const Policy = Dict{State,Action}

function to_policy(policy::Dict{Any,Any})::Policy
    pol = Policy()
    for (state, action) in policy
        pol[to_state(state)] = to_action(action)
    end
    return pol
end

function to_gym_pol(policy::Policy)::Dict{Any,Any}
    pol = Dict()
    for (state, action) in policy
        s = (Int(state.hr), Int(state.bp), Int(state.o2), Int(state.glu), state.diabetic, state.abx, state.vaso, state.vent)
        pol[s] = (action.abx, action.vaso, action.vent)
    end
    return pol
end

struct SepsisParams
    get_parameters::Any
    set_parameters::Function
    hr_probs::Function
    bp_probs::Function
    o2_probs::Function
    glu_probs::Function
    update::Function
end

end